libraries/chain_kv/include/b1/session/undo_stack.hpp

```diff
@@ -66,18 +76,26 @@ class undo_stack {
    /// \remarks This is the next session to be committed.
    const_variant_type bottom() const;
 
+   void open();
+   void close();
+
  private:
    int64_t                  m_revision{ 0 };
    Session*                 m_head;
    std::deque<session_type> m_sessions; // Need a deque so pointers don't become invalidated.  The session holds a
                                         // pointer to the parent internally.
+   fc::path                 m_datadir;
 };
 
 template <typename Session>
-undo_stack<Session>::undo_stack(Session& head) : m_head{ &head } {}
+undo_stack<Session>::undo_stack(Session& head, const fc::path& datadir) : m_head{ &head }, m_datadir{ datadir } {
+   open();
+}
 
 template <typename Session>
 undo_stack<Session>::~undo_stack() {
+   close();
+
    for (auto& session : m_sessions) { session.undo(); }
```

Comment:
Since `close` is draining the stack, this line is no longer needed.

Reply:
Removed.